Tools/jit/README.md

```diff
@@ -0,0 +1,46 @@
+The JIT Compiler
+================
+
+This version of CPython can be built with an experimental just-in-time compiler. While most everything you already know about building and using CPython is unchanged, you will probably need to install a compatible version of LLVM first.
+
+### Installing LLVM
+
+The JIT compiler does not require end users to install any third-party dependencies, but part of it must be *built* using LLVM. You are *not* required to build the rest of CPython using LLVM, or the even the same version of LLVM (in fact, this is uncommon).
```

Comment:
Could you perhaps put a few notes about what the blockers are for GCC support / what specifically is relied upon from LLVM?

Reply:
Looks like it might be https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110899 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83324.

Reply:
I can add some comments, yeah. In addition to an equivalent of [`__attribute__((musttail))`](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83324), we also need `llvm-readobj` (`llvm-objdump` is nice to have, too). I haven't dug into whether the non-LLVM equivalents specify the features we need. There may also be some `clang`-specific flags that we're using, but I'd have to check that.

There's also the issue of not wanting to support tons of toolchains for this just because we *can*. I think there'd need to be a demonstrable need for downloading a very-recent version of GCC instead of a very-recent version of LLVM to make it worthwhile.

Reply:
I think it's not too much of a burden to support the two most popular toolchains, considering that both of them have large ecosystems where one, or the other, is the system toolchain and preferred for any and all purposes.

"Tons" is a confusing term to me here -- I don't think that's been suggested?

If it can be made to work with GCC, then IMHO the immense mindshare of GCC merits it being given precisely the same amount of special consideration as LLVM gets.

Reply:
> I think it's not too much of a burden to support the two most popular toolchains, considering that both of them have large ecosystems where one, or the other, is the system toolchain and preferred for any and all purposes.

I'm open to supporting the GCC/Binutils toolchain if/when they becomes a reasonable drop-in for the LLVM functionality we require. But again, not just because we *can*... it should be clearly better in some way to some group of people to justify the additional maintenance burden (which is real: it includes expanding our CI/buildbot matrix, and making sure that GCC support won't block future development in some way).

I'm not yet convinced that "installing a super-recent GCC/Binutils is easier that installing a super-recent LLVM" is a great argument, at least until distros start shipping the features we require. But we can cross that bridge when we get there.

Reply:
> > I think it's not too much of a burden to support the two most popular toolchains, considering that both of them have large ecosystems where one, or the other, is the system toolchain and preferred for any and all purposes.
> 
> I'm open to supporting the GCC/Binutils toolchain if/when they becomes a reasonable drop-in for the LLVM functionality we require. But again, not just because we _can_... it should be clearly better in some way to some group of people to justify the additional maintenance burden (which is real: it includes expanding our CI/buildbot matrix, and making sure that GCC support won't block future development in some way).

The big question is what exact features is missing from GCC 12 right now? I really doubt musttail is a missing must feature since it is broken even in LLVM. the use of objdump here seems a little off since the relocations that you depend on are target specific and are elf specific. There has been no real list of must have features needed to support GCC listed anywhere. Also it should note musttail seems like a broken feature in general since means it will error out if tail calls cannot happen.

Reply:
> The big question is what exact features is missing from GCC 12 right now?

That I know of? Guaranteed tail calls. I'm updating the `README` to clarify this. There may be other issues with the rest of the JIT build (compiler flags, JSON serialization of object files, etc.), but I haven't explored them *since this can't work with GCC anyways right now*.

But this PR really isn't the place to be having this discussion. Feel free to open an issue once it's landed on `main` and we can weigh the pros and cons of adding GCC support there. But that feels super premature, and I think somebody (not me) has to actually *try* it before we can even decide if it's worth it, just to collect the list of must-needed features you're asking for.

Separately, if you feel that there are bugs in this PR (or that parts are unclear), reviews are welcome.

Reply:
> > The big question is what exact features is missing from GCC 12 right now?
> 
> That I know of? Guaranteed tail calls. I'm updating the README to clarify this. There may be other issues with the rest of the JIT build (compiler flags, JSON serialization of object files, etc.), but I haven't explored them _since this can't work with GCC anyways right now_.

Guaranteed tail calls how is it required, seems like a failure in general in the JIT really. Can you better describe on why is it is needed in the README or in this patch set? Because right now it seems like it is failure in the code to require that.
As far as JSON serialization of object files, that seems like the wrong appoarch to serialize out to a text file and then read in that in. Why not read in object files directly instead? Seems like a 2 step process makes things slower, especiall when it comes to text being much more verbose than the object file itself.
> 
> But this PR really isn't the place to be having this discussion. 

Yes it is because it is about merging in support and in fact it can improve the code quality that is being merged in.